openmc/deplete/operator.py

```diff
@@ -434,137 +271,104 @@ def _differentiate_burnable_mats(self):
 
         Parameters
         ----------
-        mat : openmc.Material
-            The material to read from
-        prev_res : Results
-            Results from a previous depletion calculation
+        helper_kwargs : dict
+            Keyword arguments for helper classes
 
         """
-        mat_id = str(mat.id)
+        reaction_rate_mode = helper_kwargs['reaction_rate_mode']
+        normalization_mode = helper_kwargs['normalization_mode']
+        fission_yield_mode = helper_kwargs['fission_yield_mode']
+        reaction_rate_opts = helper_kwargs.get('reaction_rate_opts', {})
+        fission_yield_opts = helper_kwargs.get('fission_yield_opts', {})
+
+        # Get classes to assist working with tallies
+        if reaction_rate_mode == "direct":
+            self._rate_helper = DirectReactionRateHelper(
+                self.reaction_rates.n_nuc, self.reaction_rates.n_react)
+        elif reaction_rate_mode == "flux":
+            if reaction_rate_opts is None:
+                reaction_rate_opts = {}
 
-        # Get nuclide lists from geometry and depletion results
-        depl_nuc = prev_res[-1].nuc_to_ind
-        geom_nuc_densities = mat.get_nuclide_atom_densities()
+            # Ensure energy group boundaries were specified
+            if 'energies' not in reaction_rate_opts:
+                raise ValueError(
+                    "Energy group boundaries must be specified in the "
+                    "reaction_rate_opts argument when reaction_rate_mode is"
+                    "set to 'flux'.")
 
-        # Merge lists of nuclides, with the same order for every calculation
-        geom_nuc_densities.update(depl_nuc)
+            self._rate_helper = FluxCollapseHelper(
+                self.reaction_rates.n_nuc,
+                self.reaction_rates.n_react,
+                **reaction_rate_opts
+            )
+        else:
+            raise ValueError("Invalid reaction rate mode.")
 
-        for nuclide, atom_per_bcm in geom_nuc_densities.items():
-            if nuclide in depl_nuc:
-                concentration = prev_res.get_atoms(mat_id, nuclide)[1][-1]
-                volume = prev_res[-1].volume[mat_id]
-                atom_per_cc = concentration / volume
-            else:
-                atom_per_cc = atom_per_bcm * 1.0e24
+        if normalization_mode == "fission-q":
+            self._normalization_helper = ChainFissionHelper()
+        elif normalization_mode == "energy-deposition":
+            score = "heating" if self.settings.photon_transport else "heating-local"
+            self._normalization_helper = EnergyScoreHelper(score)
+        else:
+            self._normalization_helper = SourceRateHelper()
 
-            self.number.set_atom_density(mat_id, nuclide, atom_per_cc)
+        # Select and create fission yield helper
+        fission_helper = self._fission_helpers[fission_yield_mode]
+        if fission_yield_opts is None:
+            fission_yield_opts = {}
```

Comment:
```suggestion
```

Reply:
The same comment applies as it does to this `if` statement in `FluxDepletionOperator._get_helper_classes()`